libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,173 @@
+#include "soft_timer.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "critical_section.h"
+#include "interrupt.h"
+#include "status.h"
+#include "stm32f0xx_interrupt.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_tim.h"
+
+typedef struct Timer {
+  uint32_t duration;
+  void *context;
+  bool inuse;
+  SoftTimerCallback callback;
+} Timer;
+
+static Timer s_soft_timer_elems[SOFT_TIMER_MAX_TIMERS];
+static SoftTimerID s_active_timer_id;
+static uint32_t s_active_timer_duration;
+
+void prv_start_timer(uint32_t duration_us) {
+  // Stop the timer.
+  TIM_Cmd(TIM2, DISABLE);
+
+  // Clear any pending interrupts, this will always occur in an interrupt or new timer getting
+  // started in a critical section so we need to prevent a accidental trigger if the time expires in
+  // the middle of the critical section and we started a new timer which falls before it.
+  TIM_ClearITPendingBit(TIM2, TIM_IT_Update);
+
+  // Get the clock speed of the clocks to determine the SYSCLK speed which is what TIM2 uses.
+  RCC_ClocksTypeDef clock_speeds;
+  RCC_GetClocksFreq(&clock_speeds);
+
+  // Configure each clock tick to be 1 microsecond from 0 to duration.
+  TIM_TimeBaseInitTypeDef init_struct = { .TIM_Prescaler = clock_speeds.SYSCLK_Frequency / 1000000,
+                                          .TIM_CounterMode = TIM_CounterMode_Up,
+                                          .TIM_Period = duration_us,
+                                          .TIM_ClockDivision = TIM_CKD_DIV1 };
+  s_active_timer_duration = duration_us;
+
+  // Reset the counter to 0.
+  TIM_SetCounter(TIM2, 0);
+
+  // Start the timer.
+  TIM_TimeBaseInit(TIM2, &init_struct);
+  TIM_Cmd(TIM2, ENABLE);
+}
+
+void prv_soft_timer_update(uint32_t curr_time) {
+  // Default these values to max. If there is ever a newer timer it will replace them. In the event
+  // there are no timers left the defaults will result in the module being reset to a state where no
+  // timers exist.
+  SoftTimerID min_time_id = SOFT_TIMER_MAX_TIMERS;
+  uint32_t min_duration = UINT32_MAX;
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
```

Comment:
It might be faster and more accurate to have TIM2 run continuously at 1Mhz (I believe it has a 32-bit counter) and calculate the expected counter value instead of updating the duration for each active timer. We have plenty of hardware timers. :P